src/Castle.Facilities.AspNetCore/AspNetCoreFacility.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright 2004-2018 Castle Project - http://www.castleproject.org/
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.Facilities.AspNetCore
+{
+	using System;
+	using System.Threading;
+
+	using Castle.Facilities.AspNetCore.Contributors;
+	using Castle.MicroKernel.Facilities;
+
+	using Microsoft.Extensions.DependencyInjection;
+
+	public class AspNetCoreFacility : AbstractFacility
+	{
+		internal const string IsCrossWiredIntoServiceCollectionKey = "windsor-registration-is-also-registered-in-service-collection";
+
+		public static readonly AsyncLocal<AspNetCoreFacility> Current = new AsyncLocal<AspNetCoreFacility>();
```

Comment:
@jonorossi 

Hope you don't mind the ambient state here. Most of the registration, happens in a single thread, so I think we are good. I am a bit worried about doing this as I wont be `dog fooding` this on net45.

Reply:
> I don't think it makes sense to add the AspNetCoreFacility if and only if you are doing cross-wiring, otherwise it throws if you aren't doing cross-wiring.

I hear you. We can definitely wind this back to how it was. No problem. Before we do I just want to make sure I understand this better.

> I still think this static stuff is completely unnecessary and that the ComponentRegistration should be as stateless as possible, and that the "cross wiring not configured" exception should be throw by CrossWiringComponentModelContributor.ProcessModel after the user calls Register on a container instance, at this point the CMC has access to the IKernel and easy access to AspNetCoreFacility.Services of the facility instance.

For cross wiring, the CMC needs to be installed in the first place and secondly requires a reference to the IServiceCollection. If the CMC was not installed in the first place how would throwing the helpful exception from inside it help? 

Thanks for the tip on the ProcessModel. This would clean up the static reference, I still cannot see how it guard a CrossWire registration from happening unless the CMC was installed in the first place. Does this make sense?

Reply:
> We can definitely wind this back to how it was. No problem. Before we do I just want to make sure I understand this better.

I actually meant to take it further, users always add the facility, and the facility always adds the CWCMC and when it comes across a component that is set for cross wiring it sets it up, or throws if the container isn't configured correctly, the exception will wind up the stack through container.Register.

Reply:
> I actually meant to take it further

Awesome. I like where this is going. 

The original CrossWire extension could either extend the IServiceCollection and take the IWindsorContainer as a parameter or vice versa. This was awful. Introducing the CWCMC really went far to solve this problem. It did move the validation problem somewhere else though.

> users always add the facility

What we are validating against here is whether someone added the facility or not. Now we have a facility with a configuration callback. 

> and the facility always adds the CWCMC and when it comes across a component that is set for cross wiring it sets it up

Correct, happy path. 

> or throws if the container isn't configured correctly

Let's validate against the configuration callback. Not whether the facility is registered.

>  the exception will wind up the stack through container.Register

That is exactly what we are trying to do. Like it. Thanks for the input.

*Summary*

We will move the validation down into the facility configuration extension. ie. if you call AddFacility without doing the config callback for `f => f.CrossWiresInto(IServiceCollection)` we throw helpfully. @generik0 I think this seems perfectly reasonable. We also have the documentation covered. What do you think?

Reply:
@fir3pho3nixx I think this is fine. I have begun testing. It appears to be working as expected. Will reprot more ASAP.

Reply:
@generik0 Sometimes it take me days/weeks. Don't stress. This is by far the worlds craziest PR. :)

The outstanding issue for this has been open since Feb 2, 2016. 

https://github.com/castleproject/Windsor/issues/120

Don't worry fella, take your time! :)

Reply:
@jonorossi We OK to go ahead?

Reply:
> We will move the validation down into the facility configuration extension. ie. if you call AddFacility without doing the config callback for f => f.CrossWiresInto(IServiceCollection) we throw helpfully. @generik0 I think this seems perfectly reasonable. We also have the documentation covered. What do you think?

It seems like you are agreeing with what I wrote but what I wrote isn't what you are doing, so I'm a little confused.

I thought you were going to make it so you always add the facility even if you don't use cross wiring, then the `ComponentRegistration` extension methods don't need to check any configuration because they aren't "attached" to a container yet, and only when you call `Register` with the `ComponentRegistration` does the container's configuration (i.e. that is has a `IServiceCollection` set) get checked?

Reply:
> I thought you were going to make it so you always add the facility even if you don't use cross wiring

Correct. Once we established that, I was merely asking @generik0 if we move validation around the facility being present or not down a level of checking whether the IServiceCollection is assigned or not using a CWCMC as you suggested [here](https://github.com/castleproject/Windsor/pull/394#discussion_r194979053).  Just making sure we are all on the same page. 

> then the ComponentRegistration extension methods don't need to check any configuration because they aren't "attached" to a container yet.

Yes. You gave me a back reference to the IKernel [here](https://github.com/castleproject/Windsor/pull/394#discussion_r194979053). With `CrossWiringComponentModelContributor.ProcessModel` I can get at the kernel. Facility scan for `AspNetCoreFacility` and check whether IServiceCollection is null. Done. Remove all the validation from ComponentRegistration. Get rid of AsyncLocal.

Have I missed anything?

Reply:
@fir3pho3nixx yes, I had wrongly misinterpreted your "We OK to go ahead?" with the changes were made and are we ready to go ahead and merge.

Reply:
Awesome let me get this going. Will implement it tonight. Thanks ;)